### PR TITLE
my.reddit: refactor into module that supports pushshift/gdpr

### PR DESCRIPTION
**Pull request by seanbreckenridge**
not complete yet, plan to add `reddit.gdpr`
and merge that in with the other data sources here

However, I've made lots of decisions on the
structure of this already so just wanted
to push in case anything here is too off
base before I continue working on this

Related issue with lots of discussion
on how to implement this; #89

Related to this now being a non-namespace package
this is something we discussed a bit [here](https://github.com/karlicoss/HPI/issues/89#issuecomment-817110452); in summary:

In order to keep this backwards compatible, `__init__.py` is just
`from .rexport import *`, so this should work with anything using `my/reddit.py`
currently. If someone wants gdpr/pushshift, they have to opt in by changing imports
to `from my.reddit.all` and updating their config. In the future (once promnesia has supported
`my.reddit.all` for a while, probably a few months at least), the `__init__.py` can be removed
so that this turns into a namespace package again

### CODE REVIEW

**Comment by seanbreckenridge** [comment]
Also moved the `RedditThing` from promnesia to here, since theres a bunch of Protocol's here anyways

As far as changes to promnesia, I think this should be fine?

```diff
diff --git a/src/promnesia/sources/reddit.py b/src/promnesia/sources/reddit.py
index 01ab6d0..28f1f35 100644
--- a/src/promnesia/sources/reddit.py
+++ b/src/promnesia/sources/reddit.py
@@ -10,7 +10,7 @@ from ..common import Visit, Loc, extract_urls, Results, logger
 
 def index(*, render_markdown: bool = False, renderer: Optional['RedditRenderer'] = None) -> Results:
     from . import hpi
-    from my.reddit import submissions, comments, saved, upvoted
+    from my.reddit.all import submissions, comments, saved, upvoted
 
     if renderer is not None:
         assert callable(renderer), f"{renderer} is not a callable (should be a subclass of RedditRenderer)"
@@ -164,20 +164,6 @@ class RedditRenderer:
 # todo hmm, could do similar stuff in HPI?
 import typing
 if typing.TYPE_CHECKING:
-    from my.reddit import Submission, Comment, Save, Upvote
+    # These are Protocol classes defined in HPI
+    from my.reddit.common import Submission, Comment, Save, Upvote, RedditThing
 
-
-from datetime import datetime
-if typing.TYPE_CHECKING:
-    # TODO define in rexport?
-    from typing_extensions import Protocol
-    from typing import Dict, Any
-    class RedditThing(Protocol):
-        @property
-        def raw(self) -> Dict[str, Any]: ...
-        @property
-        def url(self) -> str: ...
-        @property
-        def created(self) -> datetime: ...
-        @property
-        def text(self) -> str: ...
-
```

Do you think I should make those imports safe and import the old stuff if it fails? Like:

```
try:
	from my.reddit.all import submissions, comments, saved, upvoted
except ImportErorr:
	from my.reddit import submissions, comments, saved, upvoted  # type: ignore[misc,no-redef]
```

I guess we can do that for now and add a warning, then deprecate it in a few months?

**Comment by seanbreckenridge** [comment]
Unsure how promnesia tests against this, maybe have to do a pypi release before the PR there would succeed

This is done as far as this PR goes @karlicoss , will do GDPR at some point in the future

File: my/reddit/rexport.py [hunk 1]
```diff
@@ -5,10 +5,31 @@
...
+# hmm -- since this was previously just using
+# uconfig, we can't have this inherit from
+# uconfig.rexport in the dataclass definition here
+# since then theres no way to get old attributes
+# in the migration
+
+# need to check before we subclass
+if hasattr(uconfig, "rexport"):
+    # sigh... backwards compatability
+    uconfig = uconfig.rexport  # type: ignore[attr-defined,misc,assignment]
```
[the remark this answers is not in the document]
Comment (karlicoss) [reply my/reddit/rexport.py]: ah yeah looks like it's a way to fix it!

**Comment by karlicoss** [comment]
> Also moved the RedditThing from promnesia to here, since theres a bunch of Protocol's here anyways

yeah, I guess makes sense to move here! Maybe a good opportunity to rename too while we're at it, maybe `RedditBase` or something like that?

also perhaps other protocols in `common.py` should inherit `RedditThing`? That should work right, and will result in less duplication?

> Do you think I should make those imports safe and import the old stuff if it fails

yeah, good idea! I'd probably also check against the exception message  (since ImportError is a bit broad and might make the error misleading)

> Unsure how promnesia tests against this, maybe have to do a pypi release after this is merged before the PR 

Yeah.. at the moment it's running against pypi version of HPI. When I want to test against some HPI change, I usually mess with these lines https://github.com/karlicoss/promnesia/blob/281db5dfc35a3cb0f563686e1838088a621f69c2/setup.py#L84-L87

I think ideally would be nice to have promnesia CI running both against HPI in pypi and HPI from master, but so far haven't set it up.

**Comment by seanbreckenridge** [comment]
> also perhaps other protocols in common.py should inherit RedditThing? That should work right, and will result in less duplication?

I would've thought so yeah, but when I tried it ended up with:

```diff
diff --git a/my/reddit/common.py b/my/reddit/common.py
index 46634ca..ff4a0f2 100644
--- a/my/reddit/common.py
+++ b/my/reddit/common.py
@@ -21,7 +21,7 @@ else:
 
 
 # common fields across all the Protocol classes, so generic code can be written
-class RedditThing(Protocol):
+class RedditBase(Protocol):
     @property
     def raw(self) -> Json: ...
     @property
@@ -35,62 +35,23 @@ class RedditThing(Protocol):
 
 
 # Note: doesn't include GDPR Save's since they don't have the same metadata
-class Save(Protocol):
-    @property
-    def raw(self) -> Json: ...
-    @property
-    def created(self) -> datetime_aware: ...
-    @property
-    def id(self) -> str: ...
-    @property
-    def url(self) -> str: ...
-    @property
-    def text(self) -> str: ...
+class Save(RedditBase):
     @property
     def subreddit(self) -> str: ...
 
 # Note: doesn't include GDPR Upvote's since they don't have the same metadata
-class Upvote(Protocol):
-    @property
-    def raw(self) -> Json: ...
-    @property
-    def id(self) -> str: ...
-    @property
-    def created(self) -> datetime_aware: ...
-    @property
-    def url(self) -> str: ...
-    @property
-    def text(self) -> str: ...
+class Upvote(RedditBase):
     @property
     def title(self) -> str: ...
 
 
 # From rexport, pushshift and the reddit GDPR export
-class Comment(Protocol):
-    @property
-    def raw(self) -> Json: ...
-    @property
-    def id(self) -> str: ...
-    @property
-    def created(self) -> datetime_aware: ...
-    @property
-    def url(self) -> str: ...
-    @property
-    def text(self) -> str: ...
+class Comment(RedditBase):
+    pass
 
 
 # From rexport and the GDPR export
-class Submission(Protocol):
-    @property
-    def raw(self) -> Json: ...
-    @property
-    def id(self) -> str: ...
-    @property
-    def created(self) -> datetime_aware: ...
-    @property
-    def url(self) -> str: ...
-    @property
-    def text(self) -> str: ...
+class Submission(RedditBase):
     @property
     def title(self) -> str: ...
```

```
my/reddit/all.py: note: In function "_rexport_comments":
my/reddit/all.py:18: error: Incompatible types in "yield from" (actual type "rexport.dal.Comment", expected type
"my.reddit.common.Comment")  [misc]
        yield from rexport.comments()
        ^
my/reddit/all.py: note: In function "_rexport_submissions":
my/reddit/all.py:23: error: Incompatible types in "yield from" (actual type "rexport.dal.Submission", expected type
"my.reddit.common.Submission")  [misc]
        yield from rexport.submissions()
        ^
my/reddit/all.py: note: In function "_rexport_saved":
my/reddit/all.py:28: error: Incompatible types in "yield from" (actual type "rexport.dal.Save", expected type
"my.reddit.common.Save")  [misc]
        yield from rexport.saved()
        ^
my/reddit/all.py: note: In function "_rexport_upvoted":
my/reddit/all.py:33: error: Incompatible types in "yield from" (actual type "rexport.dal.Upvote", expected type
"my.reddit.common.Upvote")  [misc]
        yield from rexport.upvoted()
        ^
my/reddit/all.py: note: In function "_pushshift_comments":
my/reddit/all.py:38: error: Incompatible types in "yield from" (actual type "PComment", expected type "Comment")  [misc]
        yield from pcomments()
        ^
Found 5 errors in 1 file (checked 5 source files)
```

For whatever sub-protocols (subclasses) don't work in that way, which after some digging [seems to be part of the rejected ideas on the PEP](https://www.python.org/dev/peps/pep-0544/#toc-entry-28)

Will rename it to `RedditBase`, sure

> since ImportError is a bit broad and might make the error misleading

Ok, will bring this up in the promnesia PR when I create it

**Comment by seanbreckenridge** [comment]
In relation to the no-subclassing, I also thought and tried abstract base classes, but that isn't as extendible when it comes to new sources, and requires that someone subclass the class specifically -- goes a bit against the extensibility argument

**Comment by seanbreckenridge** [comment]
Could perhaps also be somewhere where [typeguards](https://www.python.org/dev/peps/pep-0647/) could be used? That is mostly runtime though and would result in a slight performance hit to check if something matches the `RedditBase`

**Comment by karlicoss** [comment]
> For whatever sub-protocols (subclasses?) don't work in that way, which after some digging seems to be part of the rejected ideas on the PEP

huh, indeed. However [this](https://www.python.org/dev/peps/pep-0544/#merging-and-extending-protocols) suggests it should work if you inherit subprotocols from `Protocol`?

```
from typing import Protocol

class AP(Protocol):
    a: int

class BP(AP):
    b: int

class C:
    a = 1
    b = 2

a: AP = C()
b: BP = C()
```

^ this doesn't work: 

```
a.py:14: error: Incompatible types in assignment (expression has type "C", variable has type "BP")  [assignment]
    b: BP = C()
```

but if you change B to: 

```
class BP(Protocol, AP):
    b: int
```

it passes!

**Comment by seanbreckenridge** [comment]
Yep, that does seem to work, and works in promnesia as well when it expects the `RedditBase`

**Comment by karlicoss** [comment]
nice, thanks so much! I'm happy to merge, but it says there are conflicts, can you rebase?

**Comment by seanbreckenridge** [comment]
ah, was because of the macos ci stuff

not used to rebasing so hopefully that was right

**Comment by seanbreckenridge** [comment]
Ah, I don't know how this wasn't caught by mypy -- I should've actually run this

```
~/Repos/HPI-fork/my/reddit/common.py in <module>
     36
     37 # Note: doesn't include GDPR Save's since they don't have the same metadata
---> 38 class Save(Protocol, RedditBase):
     39     @property
     40     def subreddit(self) -> str: ...

TypeError: Cannot create a consistent method resolution
order (MRO) for bases object, RedditBase
```

when actually running it can't do that

Should I make a commit to revert that?

**Comment by karlicoss** [comment]
hmm one issue I noticed, in my hpi config I have this

```
class reddit:
    @classproperty
    def export_dir(self):
        return _get_reddit_files()

    unrelated = False
    rexport: Optional[PathIsh] = code / 'rexport'
```

(I'm keeping all this legacy garbage to 'dogfood' config migrations)

So currently it stumbles over trying to treat `rexport` as `class rexport` (this is a remnant of the times when these repositories weren't proper python packages https://github.com/karlicoss/HPI/issues/79 ). I guess other people might have a similar legacy config, so would be good to handle it carefully [here](https://github.com/karlicoss/HPI/blob/8422c6e420f5e274bd1da91710663be6429c666c/my/reddit/rexport.py#L29)? I suppose checking for `isinstance(.., (Path, str))` would be good enough?

**Comment by seanbreckenridge** [comment]
Is `rexport` a path to the repository in that situation? I guess in that case you would still expect `export_dir` or `export_path` to have the JSON data files?

**Comment by karlicoss** [comment]
yeah, hopefully it would mean that they use an old config, so can just assume else branch?

**Pull request merged**